### PR TITLE
Add option to save off .pose files for each segment

### PR DESCRIPTION
**Pull request by cleong110**

### CODE REVIEW

File: sign_language_segmentation/bin.py [hunk 1]
```diff
@@ -59,15 +60,36 @@ def predict(model, pose: Pose):
 
 def get_args():
     parser = argparse.ArgumentParser()
-    parser.add_argument('--pose', required=True, type=str, help='path to input pose file')
+    parser.add_argument('--pose', required=True, type=Path, help='path to input pose file')
     parser.add_argument('--elan', required=True, type=str, help='path to output elan file')
+    parser.add_argument('--save_crops',
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:65]: for options, please use kebab-case (`--save-crops`) for for this one i recommend the naming `--save-segments`

File: sign_language_segmentation/bin.py [hunk 2]
```diff
@@ -109,9 +131,17 @@ def main():
         eaf.add_linked_file(args.pose, mimetype="application/pose")
 
     for tier_id, segments in tiers.items():
+        # print(f"TIER: {tier_id}")s
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:134]: redundant comment

File: sign_language_segmentation/bin.py [hunk 3]
```diff
@@ -59,15 +60,36 @@ def predict(model, pose: Pose):
...
 
     return parser.parse_args()
 
+def save_pose_segments(tiers, tier_id, input_file_path):
+    # reload it without any of the processing, so we get all the original points and such. 
+    with input_file_path.open("rb") as f:
+        pose = Pose.read(f.read())
+        
+    for i, segment in enumerate(tiers[tier_id]):
+                out_path = input_file_path.parent / f"{input_file_path.stem}_{tier_id}_{i}.pose"
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:82]: indentation is strange

File: sign_language_segmentation/bin.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 from pose_format.utils.generic import pose_normalization_info, pose_hide_legs, normalize_hands_3d
 
 from sign_language_segmentation.src.utils.probs_to_segments import probs_to_segments
+from pathlib import Path
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:12]: should be on top - otherwise linter will scream

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: sign_language_segmentation/bin.py (hunk 1 above)
Comment (cleong110) [reply sign_language_segmentation/bin.py:65]: That does seem like a better name. Fixing.

File: sign_language_segmentation/bin.py [hunk 5]
```diff
@@ -59,15 +61,36 @@ def predict(model, pose: Pose):
...
+                        choices=["SENTENCE","SIGN"],
+                        help='whether to save cropped .pose files')
     parser.add_argument('--video', default=None, required=False, type=str, help='path to video file')
     parser.add_argument('--subtitles', default=None, required=False, type=str, help='path to subtitle file')
     parser.add_argument('--model', default='model_E1s-1.pth', required=False, type=str, help='path to model file')
     parser.add_argument('--no-pose-link', action='store_true', help='whether to link the pose file')
 
     return parser.parse_args()
 
+def save_pose_segments(tiers, tier_id, input_file_path):
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:77]: i would rather have get_args directly above main.
otherwise there is not reasonable continuity for the file

Comment (AmitMY) [inline sign_language_segmentation/bin.py:77]: `save_pose_segments` params should have type definitions where possible

File: sign_language_segmentation/bin.py [hunk 6]
```diff
@@ -11,6 +12,7 @@
 from sign_language_segmentation.src.utils.probs_to_segments import probs_to_segments
 
 
+
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:15]: too many lines. there should be two empty lines before a function

> Surround top-level function and class definitions with two blank lines. Method definitions inside a class are surrounded by a single blank line. Extra blank lines may be used (sparingly) to separate groups of related functions.9 Sept 2024

**Review by AmitMY — CHANGES_REQUESTED** [review]
Almost there!

File: sign_language_segmentation/bin.py [hunk 7]
```diff
@@ -111,7 +134,14 @@ def main():
     for tier_id, segments in tiers.items():
         eaf.add_tier(tier_id)
         for segment in segments:
-            eaf.add_annotation(tier_id, int(segment["start"] / fps * 1000), int(segment["end"] / fps * 1000))
+            start_frame = int(segment["start"] / fps * 1000)
```
[the remark this answers is not in the document]
Comment (AmitMY) [reply sign_language_segmentation/bin.py:137]: "start" her is a `frame number`, not `seconds`
therefore, it is `frame number/fps` (gives you the seconds) `* 1000` gives you milliseconds

**Review by AmitMY — CHANGES_REQUESTED** [review]
just types left :)

File: sign_language_segmentation/bin.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cleong110) [reply sign_language_segmentation/bin.py]: ahhhh of course, that makes sense.

File: sign_language_segmentation/bin.py (hunk 5 above)
Comment (cleong110) [reply sign_language_segmentation/bin.py:61]: Agreed, I'll add them

**Comment by cleong110** [comment]
All right, I think we're about good, though there's probably more stylistic improvements to be made.
pylint doesn't like how long main is, there's some missing output annotations, and we could refactor some functions, but I think those might jobs for another day

 ChatGPT has these suggestions: https://chatgpt.com/share/675065df-c64c-800e-b171-5486a47eb024

File: sign_language_segmentation/bin.py [hunk 8]
```diff
@@ -110,21 +130,33 @@ def main():
 
     for tier_id, segments in tiers.items():
         eaf.add_tier(tier_id)
+        frames_per_millisecond = fps*1000
         for segment in segments:
-            eaf.add_annotation(tier_id, int(segment["start"] / fps * 1000), int(segment["end"] / fps * 1000))
+            # convert frame numbers to millisecond timestamps, for Elan
+            start_frame_time = int(segment["start"] / fps * frames_per_millisecond)
```
Comment (AmitMY) [inline sign_language_segmentation/bin.py:136]: this implementation is now wrong. 
Was:
```py
segment["start"] / fps * 1000
```
Now:
```py
segment["start"] / fps * fps * 1000
```

**Review by AmitMY — CHANGES_REQUESTED** [review]

File: sign_language_segmentation/bin.py (hunk 8 above)
Comment (AmitMY) [reply sign_language_segmentation/bin.py:136]: I suggest 
```py
start_time = int(segment["start"] / fps * 1000)
```

Comment (cleong110) [reply sign_language_segmentation/bin.py]: Ah, it was wrong indeed! Fixed now

**Review by AmitMY — APPROVED** [review]

**Pull request merged**